app/frontend/actions.js

```diff
@@ -864,6 +937,70 @@ module.exports = ({setState, getState}) => {
     validateSendFormAndCalculateFee()
   }
 
+  const updateStakePoolAddress = (state, e) => {
+    setState({
+      stakePools: state.stakePools.map((pool) => {
+        return Object.assign(
+          {},
+          pool,
+          pool.ID === parseInt(e.target.id, 10)
+            ? {
+              address: e.target.value,
+              ...fetchPoolInfo(e.target.value),
+            }
+            : {}
+        )
+      }),
+    })
+    validateDelegationAndCalculateDelegationFee()
+  }
+
+  const updateStakePoolPercent = (state, e) => {
+    const ID = parseInt(e.target.id, 10)
+    const oldPercent = state.stakePools
+      .map((pool) => (pool.ID === ID ? pool.percent : 0))
+      .reduce((x, y) => x + y, 0)
+    const nextPercent = Math.min(
+      100 - state.delegatedPercent + oldPercent,
+      parseInt(e.target.value ? e.target.value : 0, 10)
+    )
+
+    setState({
+      delegatedPercent: state.delegatedPercent - oldPercent + nextPercent,
+      stakePools: state.stakePools.map((pool) => {
+        return Object.assign({}, pool, pool.ID === ID ? {percent: nextPercent} : {})
+      }),
+    })
+    validateDelegationAndCalculateDelegationFee()
+  }
+
+  const addStakePool = (state, e) => {
+    setState({
+      stakePools: state.stakePools.concat({
+        address: '',
+        percent: 0,
+        name: '',
+        valid: false,
+        ID: state.stakePoolID + 1,
+      }),
+      stakePoolID: state.stakePoolID + 1,
+    })
+    validateDelegationAndCalculateDelegationFee()
+  }
+
+  const removeStakePool = (state, e) => {
+    const ID = parseInt(e.target.id, 10)
+    setState({
+      delegatedPercent: state.stakePools
+        .map((pool) => (pool.ID === ID ? 0 : pool.percent))
```

Comment:
why would we do it in such a weird way. Actually, we do we even store `delegatedPercent` in state at all if we can calculate it?